### DISCUSSION TITLE
How to Lazy Load Async

### DISCUSSION THREAD

**Discussion by saulyt** [MarimerLLC/csla#2094] [Questions] [answered]
Can anyone point me to, or provide a sample of lazy loading a  child list property asynchronously. I believe the one in the book is outdated. I tried using LazyGetPropertyAsync in the getter but I am getting an error  'Property get not allowed'. I have no Business Rules attached to the property and the exception is not of type Csla.Security.SecurityException. 

It looks to me that the error is coming from this line of code in BusinessBase ReadPropery method:
```
if (((propertyInfo.RelationshipType & RelationshipTypes.LazyLoad) == RelationshipTypes.LazyLoad) && !FieldManager.FieldExists(propertyInfo))
            throw new InvalidOperationException(Resources.PropertyGetNotAllowed);
```
My code looks like this:

```
        Func<int, int,Task<FileLinkList>> FileLinkListFunction = async (cid, lid) => await DataPortal.FetchAsync<FileLinkList>(cid, lid);
        public static readonly PropertyInfo<FileLinkList> AttachmentsProperty = RegisterProperty<FileLinkList>(nameof(Attachments), RelationshipTypes.LazyLoad);
        public FileLinkList Attachments
        {
            get
            {
                return LazyGetPropertyAsync(AttachmentsProperty, FileLinkListFunction(CorrespondenceID, LinkTypeID));
            }
            private set
            {
                LoadProperty(AttachmentsProperty, value);
            }
        }
```

Thank you

**Comment by saulyt** [comment, marked as answer]
I tried removing the RelationshipTypes.LazyLoad from the RegisterProperty functions. This eliminated the error but the fetch is happening every time the property is used. It is not caching the result.

**Reply by rockfordlhotka** [reply]
It should not be necessary to use `RelationshipTypes` for this purpose.

What is happening is that C# evaluates the parameters of the `LazyGetPropertyAsync` method before calling the method itself. So it is C# that is invoking your `FileLinkListFunction` method so it can pass the result to `LazyGetPropertyAsync` (which ignores the value after the first call, because it already has the value).

The solution is to provide a factory that only calls your async load method if the property value doesn't already exist. This can be done by using the `FieldExists` method:

```c#
    public static readonly PropertyInfo<PersonEdit> ChildProperty = RegisterProperty<PersonEdit>(nameof(Child));
    public PersonEdit Child
    {
      get => LazyGetPropertyAsync(ChildProperty, 
        FieldManager.FieldExists(ChildProperty) ? Task.FromResult(default(PersonEdit)) : DataPortal.FetchAsync<PersonEdit>("c"));
      set => SetProperty(ChildProperty, value);
    }
```
If the field already exists a null (in this case) is provided (because it is ignored anyway), otherwise the data portal is invoked to get the object.

**Reply by saulyt** [reply]
Thank you Rocky! That works.
